src/builder.rs

```diff
@@ -426,115 +456,149 @@ impl SaplingBuilder {
-        indexed_outputs.shuffle(&mut rng);
-
-        // Record the transaction metadata and create dummy outputs.
-        let spend_infos = indexed_spends
-            .into_iter()
-            .enumerate()
-            .map(|(i, (pos, spend))| {
-                // Record the post-randomized spend location
-                tx_metadata.spend_indices[pos] = i;
+        bundle::<SP, OP, _, _>(
+            rng,
+            self.spends,
+            self.outputs,
+            self.bundle_type,
+            self.zip212_enforcement,
+        )
+    }
+}
 
-                spend
-            })
-            .collect::<Vec<_>>();
-        let output_infos = indexed_outputs
-            .into_iter()
-            .enumerate()
-            .map(|(i, output)| {
-                if let Some((pos, output)) = output {
-                    // Record the post-randomized output location
-                    tx_metadata.output_indices[pos] = i;
-
-                    output
-                } else {
-                    // This is a dummy output
-                    SaplingOutputInfo::dummy(&mut rng, self.zip212_enforcement)
-                }
-            })
-            .collect::<Vec<_>>();
+/// Constructs a new Sapling transaction bundle of the given type from the specified set of spends
+/// and outputs.
+pub fn bundle<SP: SpendProver, OP: OutputProver, R: RngCore, V: TryFrom<i64>>(
+    mut rng: R,
+    spends: Vec<SpendInfo>,
+    outputs: Vec<OutputInfo>,
+    bundle_type: BundleType,
+    zip212_enforcement: Zip212Enforcement,
+) -> Result<Option<(UnauthorizedBundle<V>, SaplingMetadata)>, Error> {
+    if let BundleType::Transactional { anchor } = bundle_type {
+        for spend in &spends {
+            if !spend.has_matching_anchor(anchor) {
+                return Err(Error::AnchorMismatch);
+            }
+        }
+    } else if !spends.is_empty() {
+        return Err(Error::BundleTypeNotSatisfiable);
+    }
 
-        // Compute the transaction binding signing key.
-        let bsk = {
-            let spends: TrapdoorSum = spend_infos.iter().map(|spend| &spend.rcv).sum();
-            let outputs: TrapdoorSum = output_infos.iter().map(|output| &output.rcv).sum();
-            (spends - outputs).into_bsk()
-        };
+    let requested_output_count = outputs.len();
+    let bundle_output_count = bundle_type
+        .num_outputs(spends.len(), requested_output_count)
+        .map_err(|_| Error::BundleTypeNotSatisfiable)?;
+
+    // Record initial positions of spends and outputs
+    let mut indexed_spends: Vec<_> = spends
+        .into_iter()
+        .enumerate()
+        .map(|(i, s)| (i, s.prepare(&mut rng)))
+        .collect();
+
+    let padding_outputs = iter::repeat_with(|| OutputInfo::dummy(&mut rng))
+        .take(bundle_output_count - requested_output_count)
+        .collect::<Vec<_>>();
+    let mut indexed_outputs: Vec<_> = outputs
+        .into_iter()
+        .chain(padding_outputs.into_iter())
```

Comment:
This alters the `SaplingMetadata` to include indices that are "out of range" (do not correspond to `Builder::add_output` calls, or entries in `outputs`. This could cause issues in downstream code, so we should preserve the existing semantics of `SaplingMetadata`.

(In the `orchard` crate we never added this, instead requiring the caller to trial-decrypt themselves to figure out where their notes went; we could consider removing `SaplingMetadata` from the builder API and requiring the same here, but that should be a separate PR.)